crates/jstzd/src/protocol/mod.rs

```diff
@@ -0,0 +1,186 @@
+use rust_embed::Embed;
+use serde_json::Value;
+use std::fmt::Display;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ProtocolConstants {
+    Sandbox,
+    #[cfg(test)]
+    TestConstants,
+}
+
+impl Default for ProtocolConstants {
+    fn default() -> Self {
+        Self::Sandbox
+    }
+}
+
+impl Display for ProtocolConstants {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(match self {
+            Self::Sandbox => "sandbox",
+            #[cfg(test)]
+            Self::TestConstants => "test-test",
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum Protocol {
+    Alpha,
+    #[cfg(test)]
+    TestVersion,
+}
+
+impl Default for Protocol {
+    fn default() -> Self {
+        Self::Alpha
+    }
+}
+
+impl Protocol {
+    fn hash(&self) -> &'static str {
+        match self {
+            Protocol::Alpha => "ProtoALphaALphaALphaALphaALphaALphaALphaALphaDdp3zK",
+            #[cfg(test)]
+            Protocol::TestVersion => "test-hash",
+        }
+    }
+
+    fn parameter_file(&self, constants: &ProtocolConstants) -> PathBuf {
+        Path::new(&constants.to_string()).join(self.hash())
+    }
+}
+
+#[cfg(not(test))]
+#[derive(Embed)]
+#[folder = "$CARGO_MANIFEST_DIR/resources/protocol_parameters/"]
+pub struct ProtocolParameterFile;
+
+#[cfg(test)]
+#[derive(Embed)]
+#[folder = "$CARGO_MANIFEST_DIR/resources/test/protocol_parameters/"]
+pub struct ProtocolParameterFile;
+
+#[derive(Default)]
+pub struct ProtocolParameterBuilder {
+    protocol: Protocol,
+    constants: ProtocolConstants,
+    source_path: Option<PathBuf>,
+}
+
+impl ProtocolParameterBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn set_protocol(&mut self, protocol: Protocol) -> &mut Self {
+        self.protocol = protocol;
+        self
+    }
+
+    pub fn set_constants(&mut self, constants: ProtocolConstants) -> &mut Self {
+        self.constants = constants;
+        self
+    }
+
+    pub fn set_source_path(&mut self, path: &str) -> &mut Self {
+        self.source_path = Some(PathBuf::from(path));
+        self
+    }
+
+    pub fn build(&mut self) -> anyhow::Result<PathBuf> {
+        let mut raw_json = self.load_parameter_json()?;
+        let json = raw_json.as_object_mut().ok_or(anyhow::anyhow!(
+            "Failed to convert loaded json file into a json object"
+        ))?;
+
+        let output_path = tempfile::NamedTempFile::new().unwrap().path().to_path_buf();
+        let file = std::fs::File::create(&output_path)?;
```

Comment:
Hmm `tempfile::NamedTempFile::new()` should be enough to create the file but I guess we don't want to drop the temp file when the function exits. Instead, I suggest adding the following trait marker 
```
pub trait ReadWritable: Read + Write {}

impl ReadWritable for tempfile::NamedTempFile {}
```

then this function will return a Result<impl ReadWritable> and this part can become: 
```
let file = tempfile::NamedTempFile::new().unwrap();
serde_json::to_writer(file.as_file(), &json)?;
Ok(file)
```

Reply:
Does it matter if we drop the temp file? Temp files are different from `TempDir`: they don't get removed when the instance is dropped.

Reply:
`NamedTempFile` owns `TempPath` and if it drops then `TempPath` gets drop which will delete the file.

 I think what's happening here is that when you call `to_path_buf()`, the path is copied and because there is no reference to `NamedTempFile`, it gets deleted. The file then gets created again on the next line.

Reply:
I see. Yes I realised I was wrong about that when I tried to use a temp source file in the last test case. I've updated this.